### PR TITLE
Adding jHyphen as a fallback

### PR DESCRIPTION
**Pull request by kalaspuffar**
As the current implementation in dotify doesn't support all daisy organization languages we decided to implement a fallback in order to run hyphenation with libhyphen if the language isn't supported by the built-in latex hyphenator.

### CODE REVIEW

**Comment by kalaspuffar** [comment]
Hi @bertfrees and @PaulRambags

I know that you don't use this code in the Pipeline but you might want to review it and give some feedback anyway.

Best regards
Daniel

File: src/org/daisy/dotify/hyphenator/impl/JHyphenator.java
Comment (bertfrees) [inline src/org/daisy/dotify/hyphenator/impl/JHyphenator.java]: I believe this will throw an exception if the word contains non-standard hyphenation points. You might want to handle that?

File: src/org/daisy/dotify/hyphenator/impl/JHyphenator.java
Comment (bertfrees) [inline src/org/daisy/dotify/hyphenator/impl/JHyphenator.java]: `replace` instead of `replaceAll` might be more efficient? Dunno.

**Comment by bertfrees** [comment]
If you're interested I have libhyphen binaries for Windows and MacOS.

File: src/org/daisy/dotify/hyphenator/impl/JHyphenator.java
Comment (PaulRambags) [inline src/org/daisy/dotify/hyphenator/impl/JHyphenator.java]: this line contains a mixture of TAB and space characters

File: src/org/daisy/dotify/hyphenator/impl/JHyphenator.java
Comment (PaulRambags) [inline src/org/daisy/dotify/hyphenator/impl/JHyphenator.java]: this line contains a mixture of TAB and space characters

File: src/org/daisy/dotify/hyphenator/impl/JHyphenator.java
Comment (PaulRambags) [inline src/org/daisy/dotify/hyphenator/impl/JHyphenator.java]: You are performing the same action twice because `s` consists of `start` + `rest` and `start` was already processed in the previous statement. That is not so nice.

**Review by PaulRambags — CHANGES_REQUESTED** [review]

**Comment by kalaspuffar** [comment]
Hi, @bertfrees and @PaulRambags.

I've rewritten the code a bit to improve performance and hopefully resolve some of the issues you've reported. Not much of the old code is left :)

So if you have time, please review it again.

PS. @bertfrees I've also tested a bit more, and jhyphen has pretty much the same performance if we use it as I implemented it here. I think most of the issues with time can be because of the regression test. I think they take longer if many documents differ. DS.

Best regards
Daniel

**Comment by bertfrees** [comment]
Looks good.

> I think most of the issues with time can be because of the regression test. I think they take longer if many documents differ.

Ah yes, that could be an explanation.

**Review by PaulRambags — APPROVED** [review]

**Comment by kalaspuffar** [comment]
Hi @bertfrees 

Does that mean that you approve this PR as well? When it comes to the other DLLs or so files for the jhyphen library, then we are interested in getting hold of these. We would like to test it on all systems, at least. 

Do you have any code to locate the dictionary files on different platforms?

Best regards
Daniel

**Comment by bertfrees** [comment]
You don't need my approval. But as far as I was able to tell (I didn't look into every little aspect of the change) it looked OK.

In my code I embed the binaries and install them in some location before invoking them so I don't need to locate them. I don't have a fallback to use system libraries. You can find the binaries here: https://repo1.maven.org/maven2/org/daisy/libs/hyphen/2.8.8/. They are a bit hidden inside the JARs but you'll find them. There are versions for Linux, Windows and Mac OS, and for both 32 and 64 bit architectures.

**Pull request merged**
